### ISSUE TITLE
How to do authenticated schema introspection requests?

### ISSUE THREAD

**Issue by madebyfabian** [dulnan/nuxt-graphql-middleware#10]
Hi there! Love this module, exactly what I needed to build a headless WordPress GraphQL website.
One thing though: My /graphql endpoint is only allowing authenticated schema introspection requests (which is recommended anyway). So my config is:

```ts
export default defineNuxtConfig({
  // ...
  graphqlMiddleware: {
    graphqlEndpoint: process.env.NUXT_GQL_HOST,
    serverFetchOptions() {
      return {
        headers: {
          Authorization: process.env.NUXT_GQL_TOKEN,
        },
      };
    },
    codegenConfig: {
      schema: {
        [process.env.NUXT_GQL_HOST as string]: {
          headers: {
            Authorization: process.env.NUXT_GQL_TOKEN,
          },
        },
      },
    },
  },
});
´
```

When executing this though, the CLI errors:

```bash
Failed to load schema from https://mywpurl.com/graphql:
    GraphQL introspection is not allowed, but the query contained __schema or __type
        GraphQLError: GraphQL introspection is not allowed, but the query contained __schema or __type
    at createGraphQLError (file:///Users/fabian/Desktop/code/madebyfabian/node_modules/.pnpm/@graphql-tools+utils@9.2.1_graphql@16.6.0/node_modules/@graphql-tools/utils/esm/errors.js:6:12)
    at file:///Users/fabian/Desktop/code/madebyfabian/node_modules/.pnpm/@graphql-tools+wrap@9.3.5_graphql@16.6.0/node_modules/@graphql-tools/wrap/esm/introspect.js:10:71
    at Array.map (<anonymous>)
    at getSchemaFromIntrospection (file:///Users/fabian/Desktop/code/madebyfabian/node_modules/.pnpm/@graphql-tools+wrap@9.3.5_graphql@16.6.0/node_modules/@graphql-tools/wrap/esm/introspect.js:10:58)
    at file:///Users/fabian/Desktop/code/madebyfabian/node_modules/.pnpm/@graphql-tools+wrap@9.3.5_graphql@16.6.0/node_modules/@graphql-tools/wrap/esm/introspect.js:41:32
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async UrlLoader.load (file:///Users/fabian/Desktop/code/madebyfabian/node_modules/.pnpm/@graphql-tools+url-loader@7.17.11_graphql@16.6.0/node_modules/@graphql-tools/url-loader/esm/index.js:216:29)
    at async file:///Users/fabian/Desktop/code/madebyfabian/node_modules/.pnpm/@graphql-tools+load@7.8.12_graphql@16.6.0/node_modules/@graphql-tools/load/esm/load-typedefs/load-file.js:14:39
    at async Promise.all (index 5)
    at async loadFile (file:///Users/fabian/Desktop/code/madebyfabian/node_modules/.pnpm/@graphql-tools+load@7.8.12_graphql@16.6.0/node_modules/@graphql-tools/load/esm/load-typedefs/load-file.js:12:9)
    at async file:///Users/fabian/Desktop/code/madebyfabian/node_modules/.pnpm/@graphql-tools+load@7.8.12_graphql@16.6.0/node_modules/@graphql-tools/load/esm/load-typedefs/collect-sources.js:201:25
```

Tried to use codegen without the module to make sure it works, so the working `codegen.ts` for demo purposes would be:

```ts
// codegen.ts
import type { CodegenConfig } from "@graphql-codegen/cli";
import * as dotenv from "dotenv";
dotenv.config();

const config: CodegenConfig = {
  schema: {
    [process.env.NUXT_GQL_HOST as string]: {
      headers: {
        Authorization: process.env.NUXT_GQL_TOKEN,
      },
    },
  } as any,

  documents: [
    "./server/trpc/**/*.ts",
    "./server/api/v2/**/*.ts",
    "./api/**/*.ts",
    "!./graphql/**/*",
  ],

  generates: {
    "./graphql/": {
      preset: "client",
      plugins: [],
    },
  },
};

export default config;

```

**Comment by dulnan** [comment]
Hey! I‘m glad you find this module useful!
The codegenConfig property contains the actual config, so without the `[process.env.NUXT_GQL_HOST as string]`:

```typescript
export default defineNuxtConfig({
  graphqlMiddleware: {
    codegenConfig: {
      headers: {
        Authorization: process.env.NUXT_GQL_TOKEN
      },
    },
  }
})
```

I quickly looked through the code and it seems like that config is not passed to the underlying function call that downloads the schema, only for TS codegen. I will fix that sometime today. Because the config is different to the one used for TS codegen I‘ll introduce a new property graphqlMiddleware.schemaConfig that will be used for this codegen part.

**Comment by madebyfabian** [comment]
Thanks so much!

**Comment by madebyfabian** [comment]
Just a note: I think the TypeScript types on `codegenConfig` are wrong:

<img width="589" alt="Bildschirm­foto 2023-02-19 um 09 05 47" src="https://user-images.githubusercontent.com/32813692/219936480-ff467675-3912-4eaa-b29a-648b17b05edf.png">

```bash
Type '{ headers: { Authorization: string; }; }' is not assignable to type 'TypeScriptDocumentsPluginConfig'.
  Object literal may only specify known properties, and 'headers' does not exist in type 'TypeScriptDocumentsPluginConfig'.
```

**Comment by madebyfabian** [comment]
I also discovered that `graphqlMiddleware.codegenConfig.documents` doesn't work. My usecase is that I want to exclude some `.gql` files entirely, since the schema generation fails due to an upstream bug https://github.com/dotansimha/graphql-code-generator/issues/9010

So I want to use something like 
```ts
{
  graphqlMiddleware: {
    graphqlEndpoint: process.env.NUXT_GQL_HOST,
    serverFetchOptions: {
      headers: {
        Authorization: process.env.NUXT_GQL_TOKEN || "",
      },
    },
    codegenConfig: {
      documents: [
        // Exclude due to infinite loop
        "!pages/_graphql/SinglePage.query.gql",
        "!pages/_graphql/SinglePost.query.gql",
      ],
      headers: {
        Authorization: process.env.NUXT_GQL_TOKEN || "",
      },
    } as any,
  },
};
```

also tried it with the `documents` and `autoImportPatterns` config, but regardless of which I set, the module is always trying to generate schema for those `.gql` files I don't want it to.

**Comment by dulnan** [comment]
I released a new beta that adds support for schema-ast configuration: `nuxt-graphql-middleware@3.1.0-beta.3`

For the authentication headers see this doc page here:
https://github.com/dulnan/nuxt-graphql-middleware/blob/module-server-options/docs/configuration/using-auth-headers.md

This beta also contains a breaking change I prepared two weeks ago where some configuration options that are needed during server runtime have been moved to a separate file that is inlined into the .output server build. Previously the module tried to load nuxt.config.ts at runtime, which failed if only the .output folder is deployed (which actually is the correct way).

If you use a method for `graphqlEndpoint` or the `serverFetchOptions`, `onServerError` or `onServerResponse` options, you'll need to move those to `~/app/graphqlMiddleware.serverOptions.ts`. That was the only way to make this work. See this page for an explanation: https://github.com/dulnan/nuxt-graphql-middleware/blob/module-server-options/docs/configuration/server-options.md

**Comment by madebyfabian** [comment]
@dulnan Sounds great, currently implementing it. I now get an infinite "Reloading Server" with no additional info. Did that happen to you too?
![Bildschirm­foto 2023-02-19 um 14 36 25](https://user-images.githubusercontent.com/32813692/219951712-a662070a-e165-4f79-a059-a37c506f303b.jpg)

Does that mean that everything I load into `~/app/graphqlMiddleware.serverOptions.ts` is available on the client bundle, making people able to inspect for my auth token for example?

**Comment by dulnan** [comment]
I didn't get that inside the module's playground, but Nuxt's playground setup tends to sometimes work, but then installing the module in an actual Nuxt app produces errors. Have you tried removing `.nuxt` and trying again?

The serverOptions file is not part of the client bundle, only the server bundle. This workaround is needed because Nuxt doesn't offer a way for module authors to provide module users a way to pass in methods as configuration. Runtime config doesn't work for methods because it goes through JSON.stringify. The workaround is to take the serverOptions file during build and telling Nitro to inline it in the server bundle.

Of course if you accidentally import the file in a plugin or component (which is pretty hard to do accidentally :smile: ) you could probably get its contents in a client bundle. But then again, the token is likely in an ENV variable and is not resolved to a string during build, so you should be safe even in this case.

**Comment by madebyfabian** [comment]
Hi, so I managed to remove that error, was some wierd other thing in my repo I did while upgrading.

Alright sounds great! Using the beta now and the header config for both generation and runtime query execution works. That plugin saves me so much code 😄 (mostly generated stuff to be fair): 
![Bildschirm­foto 2023-02-19 um 21 12 05](https://user-images.githubusercontent.com/32813692/219972793-f414fb8a-6b47-49f4-bc74-1751d619628a.jpg)

By the way, the generated `schema.graphql` should be commited to git or not? Since it's generated on every build/dev start.
